gctree/phylip_parse.py

```diff
@@ -127,28 +127,72 @@ def parse_outfile(outfile, abundance_file=None, root="root"):
+        for site in range(len(tree.sequence)):
             base = node.sequence[site]
-            if base not in "ACGT-":
-                new_base = random.choice(ambiguous_dna_values[base])
-                for node2 in node.traverse(
-                    is_leaf_fn=lambda n: False
-                    if base in [n2.sequence[site] for n2 in n.children]
-                    else True
-                ):
-                    if node2.sequence[site] == base:
-                        node2.sequence = (
-                            node2.sequence[:site]
-                            + new_base
-                            + node2.sequence[(site + 1) :]
-                        )
+            if base not in bases:
+
+                def is_leaf(node):
+                    return (node.is_leaf()) or (node.sequence[site] in bases)
+
+                # First pass of Sankoff: compute cost vectors
+                for node2 in node.traverse(strategy="postorder", is_leaf_fn=is_leaf):
+                    base2 = node2.sequence[site]
+                    node2.cv = code_vectors[base2].copy()
+                    if not is_leaf(node2):
+                        for i in range(5):
+                            for child in node2.children:
+                                node2.cv[i] += min(
+                                    [
+                                        sum(v)
+                                        for v in zip(child.cv, cost_adjust[bases[i]])
+                                    ]
+                                )
+                # Second pass: Choose base and adjust children's cost vectors
+                if not node.is_root():
+                    node.cv = [
+                        sum(v)
+                        for v in zip(node.cv, cost_adjust[node.up.sequence[site]])
+                    ]
+                # traverse evaluates is_leaf(node) after yielding node.
+                # Resolving base makes is_leaf true; must get order before
+                # making changes.
+                preorder = list(node.traverse(strategy="preorder", is_leaf_fn=is_leaf))
+                for node2 in preorder:
+                    if node2.sequence[site] in bases:
+                        continue
+                    min_cost = min(node2.cv)
+                    base_index = random.choice(
+                        [i for i, val in enumerate(node2.cv) if val == min_cost]
+                    )
+                    new_base = bases[base_index]
+                    # Adjust child cost vectors
+                    if not is_leaf(node2):
+                        for child in node2.children:
+                            child.cv = [
+                                sum(v) for v in zip(child.cv, cost_adjust[new_base])
+                            ]
+                    node2.sequence = (
+                        node2.sequence[:site] + new_base + node2.sequence[(site + 1) :]
+                    )
     return tree
```

Comment:
I suggest cleaning up the cost vector node features like so
```suggestion
    for node in tree.traverse():
        node.del_feature("cv")
    return tree
```